fastlane_core/lib/fastlane_core/itunes_transporter.rb

```diff
@@ -180,9 +180,132 @@ def additional_upload_parameters
     end
   end
 
+  # Generates commands and executes the altool.
+  class AltoolTransporterExecutor < TransporterExecutor
+    ERROR_REGEX = /.*\*\*\*\sError:\s+(.+)/
+
+    private_constant :ERROR_REGEX
+
+    def execute(command, hide_output)
+      if Helper.test?
+        yield(nil) if block_given?
+        return command
+      end
+
+      @errors = []
+      @warnings = []
+      @all_lines = []
+
+      if hide_output
+        # Show a one time message instead
+        UI.success("Waiting for App Store Connect transporter to be finished.")
+        UI.success("Application Loader progress... this might take a few minutes...")
+      end
+
+      begin
+        exit_status = FastlaneCore::FastlanePty.spawn(command) do |command_stdout, command_stdin, pid|
+          begin
+            command_stdout.each do |line|
+              @all_lines << line
+              parse_line(line, hide_output) # this is where the parsing happens
+            end
+          end
+        end
+      rescue => ex
+        # FastlanePty adds exit_status on to StandardError so every error will have a status code
+        exit_status = ex.exit_status
+        @errors << ex.to_s
+      end
+
+      unless exit_status.zero?
+        @errors << "The call to the altool completed with a non-zero exit status: #{exit_status}. This indicates a failure."
+      end
+
+      if @warnings.count > 0
+        UI.important(@warnings.join("\n"))
+      end
+
+      if @errors.count > 0 && @all_lines.count > 0
+        # Print out the last 18 lines, this is key for non-verbose mode
+        @all_lines.last(18).each do |line|
+          UI.important("[altool] #{line}")
+        end
+        UI.message("Application Loader output above ^")
+        UI.error(@errors.join("\n"))
+      end
+
+      yield(@all_lines) if block_given?
+      return exit_status.zero?
+    end
+
+    def build_upload_command(username, password, source = "/tmp", provider_short_name = "", jwt = nil, platform = nil)
+      use_jwt = !jwt.to_s.empty?
+      [
+        "xcrun altool",
+        "--upload-app",
+        ("-u #{username.shellescape}" unless use_jwt),
+        ("-p #{password.shellescape}" unless use_jwt),
+        ("--appKey #{jwt}" if use_jwt),
+        "-t #{platform}",
+        file_upload_option(source),
+        additional_upload_parameters,
+        "-k 100000"
+      ].compact.join(' ')
+    end
+
+    def additional_upload_parameters
+      env_deliver_additional_params = ENV["DELIVER_ALTOOL_ADDITIONAL_UPLOAD_PARAMETERS"]
+      if env_deliver_additional_params.to_s.strip.empty?
+        return nil
+      end
```

Comment:
It will be crash. Care nil values

Reply:
if my understanding is correct, it has no safe because the function used only below code and it is filtered by `compact`.

https://github.com/fastlane/fastlane/blob/0f2a3d4e596d9b8d344a5b68270ab03d3daba0ab/fastlane_core/lib/fastlane_core/itunes_transporter.rb#L239-L249

But if you think that it is better to make caller not being care it, I agree and I can fix it as "", What do you think?

Reply:
done https://github.com/fastlane/fastlane/pull/20631/commits/d6755514d85ba28a28f2dedbf43a30670d093b28